storm-core/src/jvm/org/apache/storm/executor/spout/SpoutExecutor.java

```diff
@@ -0,0 +1,248 @@
+
+        this.backPressureEnabled = Utils.getBoolean(stormConf.get(Config.TOPOLOGY_BACKPRESSURE_ENABLE), false);
+
+        this.lastActive = new AtomicBoolean(false);
+        this.hasAckers = StormCommon.hasAckers(stormConf);
+        this.emittedCount = new MutableLong(0);
+        this.emptyEmitStreak = new MutableLong(0);
+        this.spoutThrottlingMetrics = new SpoutThrottlingMetrics();
+    }
+
+    @Override
+    public void init(final Map<Integer, Task> idToTask) {
+        LOG.info("Opening spout {}:{}", componentId, idToTask.keySet());
+        this.idToTask = idToTask;
+        this.maxSpoutPending = Utils.getInt(stormConf.get(Config.TOPOLOGY_MAX_SPOUT_PENDING), 0) * idToTask.size();
+        this.spouts = new ArrayList<>();
+        for (Task task : idToTask.values()) {
+            this.spouts.add((ISpout) task.getTaskObject());
+        }
+        this.pending = new RotatingMap<>(2, new RotatingMap.ExpiredCallback<Long, TupleInfo>() {
+            @Override
+            public void expire(Long key, TupleInfo tupleInfo) {
+                Long timeDelta = null;
+                if (tupleInfo.getTimestamp() != 0) {
+                    timeDelta = Time.deltaMs(tupleInfo.getTimestamp());
+                }
+                failSpoutMsg(SpoutExecutor.this, idToTask.get(tupleInfo.getTaskId()), timeDelta, tupleInfo, "TIMEOUT");
+            }
+        });
+
+        this.spoutThrottlingMetrics.registerAll(stormConf, idToTask.values().iterator().next().getUserContext());
+        this.outputCollectors = new ArrayList<>();
+        for (Map.Entry<Integer, Task> entry : idToTask.entrySet()) {
+            Task taskData = entry.getValue();
+            ISpout spoutObject = (ISpout) taskData.getTaskObject();
+            SpoutOutputCollectorImpl spoutOutputCollector = new SpoutOutputCollectorImpl(
+                    spoutObject, this, taskData, entry.getKey(), emittedCount,
+                    hasAckers, rand, isEventLoggers, isDebug, pending);
+            SpoutOutputCollector outputCollector = new SpoutOutputCollector(spoutOutputCollector);
+            this.outputCollectors.add(outputCollector);
+
+            taskData.getBuiltInMetrics().registerAll(stormConf, taskData.getUserContext());
+            Map<String, DisruptorQueue> map = ImmutableMap.of("sendqueue", transferQueue, "receive", receiveQueue);
+            BuiltinMetricsUtil.registerQueueMetrics(map, stormConf, taskData.getUserContext());
+
+            if (spoutObject instanceof ICredentialsListener) {
+                ((ICredentialsListener) spoutObject).setCredentials(credentials);
+            }
+            spoutObject.open(stormConf, taskData.getUserContext(), outputCollector);
+        }
+        openOrPrepareWasCalled.set(true);
+        LOG.info("Opened spout {}:{}", componentId, idToTask.keySet());
+        setupMetrics();
+    }
+
+    @Override
+    public Object call() throws Exception {
+        while (!stormActive.get()) {
```

Comment:
storm-active check was used for waiting for activation before calling open. That's also comment on that so it is intentional, and we need to follow on that.

https://github.com/apache/storm/blob/master/storm-core/src/clj/org/apache/storm/daemon/executor.clj#L516

Same on Bolt.

I agree we can change functional style to object oriented style but this is focused to port so I'd feel more safe if we don't change the behavior and flow in this PR.